File: pepcoding/java/L2/ArrayStrings/maximumSwap.java
import java.util.*;

public class maximumSwap {
    private static void swap(char[] arr, int left, int right){
        
        char temp = arr[left];
        arr[left] = arr[right];
        arr[right] = temp;
        
    }
  // ~~~~~~~~~~User Section~~~~~~~~~~~
  public static String maximumSwap(String num) {
    // write your code here
    char[] arr = num.toCharArray();
    int[] lastIndexOfDIgit = new int[10];

    for (int i = 0; i < arr.length; i++) {
      lastIndexOfDIgit[arr[i] - '0'] = i;
    }

    for (int i = 0; i < arr.length; i++) {
      int digit = arr[i] - '0';
      int idx = i;
      for (int j = 9; j > digit; j--) {
        if (lastIndexOfDIgit[j] > i) {
          idx = lastIndexOfDIgit[j];
          break;
        }
      }

      if (idx != i) {
        swap(arr, i, idx);
        break;
      }
    }
    return String.valueOf(arr);
  }

  // ~~~~~~~~Input Management~~~~~~~~~
  public static void main(String[] args) {
    String str = "7236";
    String ans = maximumSwap(str);
    System.out.println(ans);
  }
}

File: pepcoding/scala/L2/ArrayStrings/maximumSwap.scala
import scala.util.control._

object maximumSwap {
    def swap(arr: Array[Char], left: Int, right:Int): Unit ={
        val temp = arr(left)
        arr(left) = arr(right)
        arr(right) = temp
    }
    def maximumSwap(num: Int): Int = {
        val arr: Array[Char] = num.toCharArray()
        val lastIndexOfDIgit: Array[Int] = new Array(10)
        n = arr.size-1

        for (i <- 0 to n) {
            lastIndexOfDIgit(arr(i)-'0') = i
        }

        for (i <- 0 to n) {
            val digit = arr(i) - '0';
            val idx = i;
            for (j <- (9 to digit).reverse()) {
                if (lastIndexOfDIgit(j) > i) {
                    idx = lastIndexOfDIgit(j)
                    break()
                }
            }

        if (idx != i) {
            swap(arr, i, idx);
            break;
        }
        }
    }

    def main(args: Array[String]): Unit = {
        print(maximumSwap("2798"))
    }
}
